clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
Yeah I have talked with both @onurkaraman and @becketqin. They thinks throwing away the fetched data is fine and simple. But this is a long thread and I haven't discussed other details with them.

BTW, I am not going to use initial approach, i.e. throw away data and have the loop. I am going to try the last proposal I described, which I copy & past below. This should keep the fetch response handling the same as it is in this patch.

In the `initFetches()`, we only create `fetchRequest` for those partitions whose fetched position == consumed position. By doing so, `initFetch()` will only issue `fetchRequest` for partitions which caused `OffsetOutOfRangeException`; `poll(timeout)` will unblock upon `fetchResponse`; and the next successful `fetchedRecords()` will return data from all fetched partitions since the last successful `fetchedRecords()`.

Though I think it belongs to a difference patch, I will apply the change you described above. Let's get it committed this altogether if it gets +1 from committer.